plugins/inputs/linux_cpu/linux_cpu.go

```diff
@@ -0,0 +1,226 @@
+		tags := map[string]string{"cpu": cpu.id}
+
+		failed := false
+		for name, propPath := range cpu.props {
+			v, err := readUintFromFile(propPath)
+			if err != nil {
+				acc.AddError(err)
+				failed = true
+				break
+			}
+
+			fields[name] = v
+		}
+
+		if !failed {
+			acc.AddFields("linux_cpu", fields, tags)
+		}
+	}
+
+	return nil
+}
+
+func (g *LinuxCPU) discoverCpus() ([]cpu, error) {
+	var cpus []cpu
+
+	glob := path.Join(g.PathSysfs, "devices/system/cpu/cpu[0-9]*")
+	cpuDirs, err := filepath.Glob(glob)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(cpuDirs) == 0 {
+		return nil, fmt.Errorf("no CPUs detected at: %s", glob)
+	}
+
+	for _, dir := range cpuDirs {
+		_, cpuName := filepath.Split(dir)
+		cpuNum := strings.TrimPrefix(cpuName, "cpu")
+
+		cpu := cpu{
+			id:    cpuNum,
+			path:  dir,
+			props: make(map[string]string),
+		}
+
+		var props []prop
+
+		if choice.Contains(cpufreq, g.Metrics) {
+			props = append(props,
+				prop{name: "scaling_cur_freq", path: "cpufreq/scaling_cur_freq", optional: false},
+				prop{name: "scaling_min_freq", path: "cpufreq/scaling_min_freq", optional: false},
+				prop{name: "scaling_max_freq", path: "cpufreq/scaling_max_freq", optional: false},
+				prop{name: "cpuinfo_cur_freq", path: "cpufreq/cpuinfo_cur_freq", optional: true},
+				prop{name: "cpuinfo_min_freq", path: "cpufreq/cpuinfo_min_freq", optional: true},
+				prop{name: "cpuinfo_max_freq", path: "cpufreq/cpuinfo_max_freq", optional: true},
+			)
+		}
+
+		if choice.Contains(thermal, g.Metrics) {
+			props = append(
+				props,
+				prop{name: "throttle_count", path: "thermal_throttle/core_throttle_count", optional: false},
+				prop{name: "throttle_max_time", path: "thermal_throttle/core_throttle_max_time_ms", optional: false},
+				prop{name: "throttle_total_time", path: "thermal_throttle/core_throttle_total_time_ms", optional: false},
+			)
+		}
+
+		var failed = false
+		for _, prop := range props {
+			propPath := filepath.Join(dir, prop.path)
+			err := validatePath(propPath)
+			if err != nil {
+				if prop.optional {
+					continue
+				}
+
+				g.Log.Warnf("Failed to load property %s: %s", filepath.Join(dir, prop.path), err)
+				failed = true
+				break
+			}
+
+			cpu.props[prop.name] = propPath
+		}
+
+		if len(cpu.props) == 0 {
+			g.Log.Warnf("No properties enabled for CPU %s", cpuNum)
+			failed = true
+		}
+
+		if !failed {
+			cpus = append(cpus, cpu)
+		}
+	}
+	return cpus, nil
+}
+
+func init() {
+	inputs.Add("linux_cpu", func() telegraf.Input {
+		return &LinuxCPU{
+			Metrics: []string{"cpufreq", "thermal"},
+		}
+	})
+}
+
+func validatePath(propPath string) error {
+	f, err := os.Open(propPath)
+
+	if os.IsNotExist(err) {
+		return fmt.Errorf("CPU property does not exist: [%s]", propPath)
+	}
+
+	if err != nil {
+		return fmt.Errorf("cannot get system information for CPU property: [%s] - %v", propPath, err)
+	}
+
+	_ = f.Close() // File is not written to, closing should be safe
```

Comment:
Maybe this would be nicer?

```
func validatePath(propPath string) error {
	_, err := os.Lstat(propPath)
	if err != nil {
		if os.IsNotExist(err) {
			return fmt.Errorf("CPU property does not exist: [%s]", propPath)
		}
		return fmt.Errorf("cannot get system information for CPU property: [%s] - %v", propPath, err)
	}

	return nil
}
```

Reply:
This relates to the comment from @antst. Checking whether the file exists is not enough, we need to verify that it is readable as well.

Reply:
Ok.